server/src/main/java/org/candlepin/controller/CandlepinPoolManager.java

```diff
@@ -1227,36 +1232,163 @@ void removeEntitlement(Entitlement entitlement,
         sink.queueEvent(event);
     }
 
+
+    /**
+     * Revokes the given set of entitlements.
+     *
+     * @param entsToRevoke entitlements to revoke
+     */
     @Override
     @Transactional
-    public void revokeEntitlement(Entitlement entitlement) {
-        removeEntitlement(entitlement, true);
+    public void revokeEntitlements(List<Entitlement> entsToRevoke) {
+        if (entsToRevoke.isEmpty()) {
+            return;
+        }
+
+        List<Pool> poolsToLock = new ArrayList<Pool>();
+        for (Entitlement ent: entsToRevoke) {
+            poolsToLock.add(ent.getPool());
+        }
+
+        poolCurator.lock(poolsToLock);
+
+        log.info("Batch revoking entitlements: " + entsToRevoke.size());
+
+        entsToRevoke =  new ArrayList<Entitlement>(entsToRevoke);
+
+        List<Pool> poolsToDelete = poolCurator.listBySourceEntitlements(entsToRevoke);
+
+        for (Pool pool : poolsToDelete) {
+            entsToRevoke.addAll(pool.getEntitlements());
+        }
+
+        poolCurator.lock(poolsToDelete);
+
+        for (Entitlement ent : entsToRevoke) {
+            //We need to trigger lazy load of provided products
+            //to have access to those products later in this method.
+            ent.getPool().getProvidedProducts().size();
+            Pool pool = ent.getPool();
+            pool.setConsumed(pool.getConsumed() - ent.getQuantity());
+
```

Comment:
@nguyenfilip reminder, need to talk to you about missing
pool.getEntitlements().remove(ent);